File: backend/shared/notification_sender.py
import os, hashlib, datetime, logging, requests, math
from azure.cosmos import CosmosClient, PartitionKey, exceptions

_COSMOS_CONN = os.getenv("COSMOSDB_MARKETPLACE_CONNECTION_STRING") or os.getenv("COSMOS_CONN")
_DB_NAME = os.getenv("COSMOSDB_MARKETPLACE_DATABASE_NAME")
_CONTAINER_NAME = "notifications"

_client = None
_container = None
_container_checked = False

_logger = logging.getLogger("notification_sender")

def _init():
    global _client, _container
    if _client is None:
        if not _COSMOS_CONN or not _DB_NAME:
            raise RuntimeError("Cosmos DB settings missing")
        _client = CosmosClient.from_connection_string(_COSMOS_CONN)
        db = _client.get_database_client(_DB_NAME)
        try:
            _container = db.get_container_client(_CONTAINER_NAME)
            _container.read()
        except exceptions.CosmosResourceNotFoundError:
            _logger.warning('Creating notifications container (partition key /userId)')
            _container = db.create_container(
                id=_CONTAINER_NAME,
                partition_key=PartitionKey(path="/userId"),
                offer_throughput=400
            )
        globals()['_container'] = _container

def _ensure_container():
    # simplified: always ensures (idempotent)
    try:
        _init()
    except Exception as e:
        raise RuntimeError(f'Failed to access/create notifications container: {e}')

def _short_hash(value: str) -> str:
    return hashlib.sha256(value.encode("utf-8")).hexdigest()[:10]

def store_token(user_id: str, token: str, platform: str = "android", provider: str = "expo"):
    """Store or refresh a token. Returns (doc_id, updated_flag)."""
    _ensure_container()
    now = datetime.datetime.utcnow().isoformat()
    doc_id = f"{user_id}:{_short_hash(token)}"
    try:
        existing = _container.read_item(doc_id, partition_key=user_id)
        # Already exists – refresh timestamps/platform/provider if changed.
        changed = False
        if existing.get("platform") != platform:
            existing["platform"] = platform; changed = True
        if existing.get("provider") != provider:
            existing["provider"] = provider; changed = True
        existing["lastSeenAt"] = now
        if changed:
            existing["updatedAt"] = now
        _container.upsert_item(existing)
        _logger.info("Token refresh user=%s id=%s changed=%s", user_id, doc_id, changed)
        return (doc_id, False)
    except exceptions.CosmosResourceNotFoundError:
        doc = {
            "id": doc_id,
            "userId": user_id,
            "token": token,
            "platform": platform,
            "provider": provider or "expo",
            "valid": True,
            "createdAt": now,
            "lastSeenAt": now
        }
        _container.upsert_item(doc)
        _logger.info("Stored new token doc_id=%s user=%s provider=%s", doc_id, user_id, provider)
        return (doc_id, True)

def fetch_valid_tokens(user_ids):
    """Return list of token strings for provided user ids."""
    if not user_ids:
        return []
    _ensure_container()
    tokens = []
    for uid in user_ids:
        query = {
            "query": "SELECT c.token FROM c WHERE c.userId = @u AND c.valid = true",
            "parameters": [{"name": "@u", "value": uid}]
        }
        for item in _container.query_items(query, partition_key=uid):
            if item.get("token"):
                tokens.append(item["token"])
    return tokens

EXPO_PUSH_URL = "https://exp.host/--/api/v2/push/send"
_MAX_CHUNK = 100

def send_expo_push(messages, log=None):
    """Send a list of expo push messages. Returns dict stats & invalid tokens."""
    if not messages:
        return {"sent": 0, "invalid": 0, "details": []}
    results = []
    invalid_tokens = set()
    for i in range(0, len(messages), _MAX_CHUNK):
        chunk = messages[i:i+_MAX_CHUNK]
        try:
            resp = requests.post(EXPO_PUSH_URL, json=chunk, timeout=10)
            data = resp.json()
            tickets = data.get("data") if isinstance(data, dict) else None
            if isinstance(tickets, list):
                for msg, ticket in zip(chunk, tickets):
                    status = ticket.get("status")
                    if status != "ok":
                        err = ticket.get("message")
                        details = ticket.get("details", {})
                        if log: log(f"expo_push_fail token={msg.get('to')[:16]} status={status} err={err} details={details}")
                        if details.get("error") in ("DeviceNotRegistered", "InvalidCredentials"):
                            invalid_tokens.add(msg.get("to"))
            results.append({"chunk": math.floor(i/_MAX_CHUNK)+1, "count": len(chunk), "http": resp.status_code})
        except Exception as e:
            if log: log(f"expo_push_exception {e}")
    # Mark invalid tokens
    if invalid_tokens:
        _ensure_container()
        for t in invalid_tokens:
            # token doc id pattern: userId:hash; need query to find docs
            query = {
                "query": "SELECT * FROM c WHERE c.token = @t",
                "parameters": [{"name": "@t", "value": t}]
            }
            for item in _container.query_items(query, enable_cross_partition_query=True):
                item["valid"] = False
                item["invalidatedAt"] = datetime.datetime.utcnow().isoformat()
                try:
                    _container.upsert_item(item)
                except Exception:
                    pass
    return {"sent": len(messages), "invalid": len(invalid_tokens), "details": results}

def send_chat_message_notifications(tokens, title, body, data, log=None):
    """High-level helper for chat message pushes (expo only)."""
    if not tokens:
        return {"expoSent": 0, "expoInvalid": 0}
    messages = [{
        "to": t,
        "title": title,
        "body": body,
        "data": data,
        "sound": "default",
        "channelId": "chat-messages",
        "priority": "high"
    } for t in tokens if t.startswith("ExponentPushToken")]
    stats = send_expo_push(messages, log=log)
    return {"expoSent": stats["sent"], "expoInvalid": stats["invalid"]}

File: backend/shared/notificationSender.js
// Backend push notifications module
const crypto = require('crypto');
const { CosmosClient } = require('@azure/cosmos');

const cosmosConn = process.env.COSMOSDB_MARKETPLACE_CONNECTION_STRING || process.env.COSMOS_CONN;
const dbName = process.env.COSMOSDB_MARKETPLACE_DATABASE_NAME;
const tokensContainerName = 'notifications';

if (!cosmosConn || !dbName) {
  throw new Error('Cosmos settings missing: COSMOSDB_MARKETPLACE_CONNECTION_STRING and COSMOSDB_MARKETPLACE_DATABASE_NAME required');
}

const cosmos = new CosmosClient(cosmosConn);
const tokensContainer = cosmos.database(dbName).container(tokensContainerName);

let containerChecked = false;

async function ensureContainer() {
  if (containerChecked) return;
  try {
    await tokensContainer.read(); // throws if not exists
  } catch (e) {
    console.error('[notificationSender] notifications container missing. Create container "notifications" with partition key /userId');
  }
  containerChecked = true;
}

function shortHash(s) {
  return crypto.createHash('sha256').update(s).digest('hex').slice(0, 10);
}

// Store token with proper provider handling
async function storeToken(userId, token, platform, provider = 'expo') {
  await ensureContainer();
  
  const id = `${userId}:${shortHash(token)}`;
  const now = new Date().toISOString();
  
  const tokenDoc = {
    id,
    userId,
    token,
    platform,
    provider, // Use the actual provider parameter
    valid: true,
    createdAt: now,
    lastSeenAt: now
  };

  try {
    await tokensContainer.items.upsert(tokenDoc);
    console.log(`Token stored for user ${userId} with provider ${provider}`);
    return { id, success: true };
  } catch (error) {
    console.error('Error storing token:', error);
    throw error;
  }
}

// Fetch valid tokens for multiple users
async function fetchValidTokens(userIds) {
  if (!userIds || !userIds.length) return [];
  
  const tokens = [];
  
  try {
    for (const userId of userIds) {
      const { resources } = await tokensContainer.items.query({
        query: 'SELECT c.token, c.id, c.provider, c.platform FROM c WHERE c.userId = @uid AND c.valid = true',
        parameters: [{ name: '@uid', value: userId }]
      }).fetchAll();
      
      resources.forEach(r => tokens.push({
        userId,
        token: r.token,
        docId: r.id,
        provider: r.provider || 'expo',
        platform: r.platform
      }));
    }
    
    console.log(`Fetched ${tokens.length} valid tokens for ${userIds.length} users`);
    return tokens;
  } catch (error) {
    console.error('Error fetching tokens:', error);
    return [];
  }
}

// Mark tokens as invalid
async function markInvalid(docIds) {
  if (!docIds || !docIds.length) return;
  
  const promises = docIds.map(async (id) => {
    try {
      const userId = id.split(':')[0];
      const { resource } = await tokensContainer.item(id, userId).read();
      
      if (resource) {
        resource.valid = false;
        resource.lastError = 'DeviceNotRegistered';
        resource.updatedAt = new Date().toISOString();
        await tokensContainer.items.upsert(resource);
        console.log(`Marked token ${id} as invalid`);
      }
    } catch (error) {
      console.error(`Error marking token ${id} as invalid:`, error);
    }
  });
  
  await Promise.allSettled(promises);
}

// Send Expo push notifications
async function sendExpoPushNotifications({ recipients, title, body, data = {}, log = console.log }) {
  if (!recipients || !recipients.length) {
    return { sent: 0, invalid: 0 };
  }

  // Filter for Expo tokens only
  const expoRecipients = recipients.filter(r => 
    r.provider === 'expo' && 
    r.token && 
    r.token.startsWith('ExponentPushToken[')
  );

  if (!expoRecipients.length) {
    log('No valid Expo tokens found');
    return { sent: 0, invalid: 0 };
  }

  // Split into chunks of 100 (Expo's limit)
  const chunks = [];
  for (let i = 0; i < expoRecipients.length; i += 100) {
    chunks.push(expoRecipients.slice(i, i + 100));
  }

  let totalSent = 0;
  const invalidIds = [];

  for (const chunk of chunks) {
    const messages = chunk.map(recipient => ({
      to: recipient.token,
      title,
      body,
      data,
      sound: 'default',
      priority: 'high',
      channelId: 'chat-messages' // Android channel
    }));

    try {
      const response = await fetch('https://exp.host/--/api/v2/push/send', {
        method: 'POST',
        headers: {
          'Accept': 'application/json',
          'Content-Type': 'application/json',
          'Accept-Encoding': 'gzip, deflate'
        },
        body: JSON.stringify(messages)
      });

      if (!response.ok) {
        throw new Error(`Expo API error: ${response.status} ${response.statusText}`);
      }

      const result = await response.json();
      const results = Array.isArray(result) ? result : result.data;

      if (results) {
        results.forEach((res, index) => {
          if (res.status === 'ok') {
            totalSent++;
          } else if (res.status === 'error') {
            log(`Expo push error for token ${chunk[index].token.slice(0, 20)}...: ${res.message}`);
            
            // Mark invalid tokens
            if (res.details?.error === 'DeviceNotRegistered' || 
                res.message?.includes('not registered')) {
              invalidIds.push(chunk[index].docId);
            }
          }
        });
      }

      log(`Sent ${totalSent} Expo notifications in chunk of ${chunk.length}`);
    } catch (error) {
      log(`Expo push error for chunk: ${error.message}`);
    }
  }

  // Mark invalid tokens
  if (invalidIds.length > 0) {
    await markInvalid(invalidIds);
  }

  return { sent: totalSent, invalid: invalidIds.length };
}

// Main function to send chat notifications
async function sendChatMessageNotificationsCombined({ tokens, title, body, data = {}, log = console.log }) {
  if (!tokens || !tokens.length) {
    return {
      expoSent: 0,
      expoInvalid: 0,
      fcmSent: 0,
      fcmInvalid: 0,
      total: 0
    };
  }

  // Separate tokens by provider
  const expoTokens = tokens.filter(t => t.provider === 'expo');
  const fcmTokens = tokens.filter(t => t.provider === 'fcm');

  log(`Sending to ${expoTokens.length} Expo tokens and ${fcmTokens.length} FCM tokens`);

  // Send Expo notifications
  const expoResult = await sendExpoPushNotifications({
    recipients: expoTokens,
    title,
    body,
    data,
    log
  });

  // FCM sending would go here if implemented
  // For now, just log FCM tokens
  if (fcmTokens.length > 0) {
    log(`FCM sending not implemented yet, skipping ${fcmTokens.length} FCM tokens`);
  }

  const totalSent = expoResult.sent;
  const totalInvalid = expoResult.invalid;

  log(`Notification results: ${totalSent} sent, ${totalInvalid} invalid tokens marked`);

  return {
    expoSent: expoResult.sent,
    expoInvalid: expoResult.invalid,
    fcmSent: 0, // Not implemented yet
    fcmInvalid: 0, // Not implemented yet
    total: totalSent
  };
}

// Helper function to update device token (for the API endpoint)
async function updateDeviceToken(req, res) {
  try {
    const { userId, email, token, platform, provider = 'expo' } = req.body;
    
    if (!token) {
      return res.status(400).json({ error: 'Token is required' });
    }

    const userIdentifier = userId || email;
    if (!userIdentifier) {
      return res.status(400).json({ error: 'userId or email is required' });
    }

    const result = await storeToken(userIdentifier, token, platform, provider);
    
    res.json({ 
      success: true, 
      message: 'Device token updated successfully',
      id: result.id
    });
  } catch (error) {
    console.error('Error updating device token:', error);
    res.status(500).json({ error: 'Failed to update device token' });
  }
}

module.exports = {
  storeToken,
  fetchValidTokens,
  sendChatMessageNotificationsCombined,
  sendExpoPushNotifications,
  markInvalid,
  updateDeviceToken
};